gatoryaml/gator_yaml.py

```diff
@@ -1,106 +1,93 @@
-        """Output a generic list item"""
-        self.output += (" " * self.spaces) * self.tabs + " -" + str(item) + "\n"
-
-    def output_key(self, key):
-        """Output a generic key"""
-        self.output += (" " * self.spaces) * self.tabs + str(key) + ":\n"
-
-    def output_key_value(self, key, value):
-        """Output a generic key and it's value"""
-        self.output += ((" " * self.spaces) * self.tabs) + str(key) + ": " + str(value) + "\n"
-
-    def is_keyword(self, key, value):
-        """Output key and value if a keyword"""
-        if key in self.keywords:
-            if key == "commits":
-                self.output += (" " * self.spaces) * self.tabs \
-                               + "--" + str(key) + " " + str(value) + "\n"
-            else:
-                self.output += (" " * self.spaces) * self.tabs + str(key) \
-                               + " " + str(value) + "\n"
-            return True
-        return False
+from .exceptions import UnexpectedValue
+
+
+def dump(header, body, indent=4, spaces=4) -> str:
+    """Input dictionary is parsed. returns string of valid YAML"""
+
+    output = parse_header(header)
+    output += parse_body(split_file_path(body), indent=indent, spaces=spaces)
+    return output
+
+
+def parse_header(header) -> str:
+    """Parses header dictionary"""
+    out = ""
+    for key, val in header.items():
+        out += str(key) + ": "
+
+        if isinstance(val, list):
+            for element in val:
+                if element != val[-1]:
+                    out += str(element) + ", "
+                else:
+                    out += str(element) + "\n"
+        else:
+            out += str(val) + "\n"
+    return out
+
+
+# pylint: disable=R0913
+# Disabled because the extra parameters are for internal use.
+def parse_body(body, output="", tabs=-1, indent=4, spaces=4, custom_keywords=None) -> str:
+    """Enumerate through the body dictionary and output each key & values.
+    If a keyword is found, special output instructions can be defined.
+    If the dictionary key has no value it will be output as a parameter."""
+    keywords = ["(pure)"]
+    if custom_keywords is not None:
+        if isinstance(custom_keywords, list):
+            keywords += custom_keywords
+        else:
+            raise Exception("Expected a list for custom_keywords, got " + str(type(custom_keywords)))
+
+    tabs += 1
+    for key in body:
+        if isinstance(body[key], dict):
+            output += output_key_header(key, tabs=tabs, spaces=spaces)
+            output = parse_body(body[key], tabs=tabs, output=output)
+        elif isinstance(body[key], list):
+            output += output_key_header(key, tabs=tabs, spaces=spaces)
+            output = print_list_body(body[key], output=output, spaces=spaces, indent=indent)
+        elif key in keywords:
+            output += str(key) + "\n"
+        elif body[key] is None or body[key] == "":
+            output += "--" + str(key) + "\n"
```

Comment:
I don't think any content (like `--`) should ever be added by this library; the user must supply the value like they want it, and we shouldn't assume they want the key as a `--thing` in a certain situation.